text/0000-offset_of.md

```diff
@@ -0,0 +1,482 @@
+[zigoffsetof]: hhttps://ziglang.org/documentation/0.9.1/#offsetOf
+[zigbitoffset]: https://ziglang.org/documentation/0.9.1/#bitOffsetOf
+[doffsetof]: https://dlang.org/spec/struct.html#struct_field_properties
+[swiftoffset]: https://github.com/apple/swift-evolution/blob/ec2028964daeda2600e49aa89fd9e59d2363433b/proposals/0210-key-path-offset.md
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+1. Should any of the features listed as "Future possibilities" be supported initially?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+This proposal is intentionally minimal, so there are a number of future
+possibilities.
+
+## Enum support (`offset_of!(SomeEnum::StructVariant, field_on_variant)`)
+
+Eventually, it may be desirable to allow `offset_of!` to access the fields
+inside the struct and tuple variants of certain enums (possibly limited to enums
+with a primitive integer representation, such as `#[repr(C)]`, `#[repr(int)]`,
+or `#[repr(C, int)]` -- where `int` is one of Rust's primitive integer types —
+u8, isize, u128, etc).
+
+For example, in the future somthing like the following could be allowed:
+
+```rs
+use core::mem::offset_of;
+
+#[repr(i8)]
+enum Event {
+    Key { pressed: bool, code: u32 },
+    Resize(u32, u32),
+}
+
+const EVENT_KEY_CODE: usize = offset_of!(Event::Key, code);
+const EVENT_KEY_PRESSED: usize = offset_of!(Event::Key, pressed);
+
+const EVENT_RESIZE_W: usize = offset_of!(Event::Resize, 0);
+const EVENT_RESIZE_H: usize = offset_of!(Event::Resize, 1);
+```
+
+In this example, the name/path of the variant is used as the first argument.
+While there are use-cases for this in low level FFI code (similar to the use
+cases for `#[repr(int)]` and `#[repr(C, int)]` enums), this may need further
+design work, and is left to the future.
+
+## Nested Field Access
+
+In C, expressions like `offsetof(struct some_struct, foo.bar.baz[3].quux)` are
+allowed, where `foo.bar.baz[3].quux` denotes a path to a derived field. This can
+be of somewhat arbitrary complexity, accessing fields of nested structs,
+performing array indexing (often this is used to access past the end of the
+array even), and so on. Similar functionality is offered by
+`MemoryLayout.offset` in Swift, where more complex language features are used to
+achieve it.
+
+This was omitted from this proposal because it is not commonly used, and can
+generally be replaced (at the cost of convenience) by multiple invocations of
+the macro.
+
+Additionally, in the future similar functionality could be added in a fully
+backwards-compatible way, either by directly allowing usage like
+`offset_of!(SomeStruct, foo.bar.baz[3].quux)`, or by requiring each field be
+comma-separated, as in `offset_of!(SomeStruct, foo, bar, baz, [3], quux)`.
+
+Note that while this example shows a combination that supports array indexing,
+it's unclear if this is actually desirable for Rust.
+
+## `memoffset::span_of!` Functionality
```

Comment:
Isn't the span easily computable using `offset_of!` and `size_of` and maybe `align_of` (if we include padding) or would this be something more subtle?

Reply:
It is if you have a `type_of` or so to compute the type of the field.

Reply:
Ah, I hadn't considered the need for that. Maybe it's a reasonable future addition.

Reply:
It turns out you can do it, it's just hacky and non-obvious. Here's a version implemented using just `memoffset::offset_of!` https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3f8a716592b64b551364105a3087b955